### ISSUE TITLE
False positive 'not-callable' for classes with customiized creation (with metaclass)

### ISSUE THREAD

**Issue by skirpichev** [PyCQA/pylint#8094]
### Bug description

I'm trying to enable the 'not-callable' checker for the [Diofant](https://github.com/diofant/diofant/) project.  Unfortunately, it seems the checker has problems with customized by metaclass classes, that define symbolic functions in my case.  From the python point of view they are perfectly callable:
```
>>> from diofant import Function, sin
>>> f = Function('f')
>>> f
f
>>> f(1)
f(1)
>>> callable(f)
True
>>> callable(sin)
True
>>> sin(1)
sin(1)
```

Sorry for not providing a minimal example yet (I'll try asap).  But the issue is easily can be reproduced (see below) and I think there are no similar issues reported.

### Command used

```shell
pylint --disable=all --enable=not-callable diofant/tests/core/test_basic.py
```

### Pylint output

```shell
************* Module diofant.tests.core.test_basic
diofant/tests/core/test_basic.py:173:37: E1102: sin(x) is not callable (not-callable)
diofant/tests/core/test_basic.py:174:37: E1102: sin(x) is not callable (not-callable)
diofant/tests/core/test_basic.py:196:11: E1102: f is not callable (not-callable)
```

### Expected behavior

No errors.

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.2
Python 3.11.1+ (heads/3.11:ba88628808, Jan 14 2023, 08:40:06) [GCC 10.2.1 20210110]
```

### Additional dependencies

See https://diofant.readthedocs.io/en/latest/install.html#from-sources

**Comment by DanielNoord** [comment]
This should be handled by a specific `diofant-pylint` plugin. It is just to dynamic for `pylint` to statically infer under normal conditions.
